gnomad/utils/sparse_mt.py

```diff
@@ -991,67 +992,91 @@ def densify_all_reference_sites(
-    :param coverage_over_x_bins: List of boundaries for computing samples over X
-    :param row_key_fields: List of row key fields to use for joining `mtds` with
-        `reference_ht`
-    :param strata_expr: Optional list of dicts containing expressions to stratify the
-        coverage stats by. Only one of `group_membership_ht` or `strata_expr` can be
-        specified.
-    :param group_membership_ht: Optional Table containing group membership annotations
-        to stratify the coverage stats by. Only one of `group_membership_ht` or
-        `strata_expr` can be specified.
-    :return: Table with per-base coverage stats.
+    :param mtds: Input sparse Matrix Table or VariantDataset.
+    :param reference_ht: Table of reference sites.
+    :param entry_agg_funcs: Dict of entry aggregation functions to perform on the
+        VariantDataset/MatrixTable. The keys of the dict are the names of the
+        annotations and the values are tuples of functions. The first function is used
+        to transform the `mt` entries in some way, and the second function is used to
+        aggregate the output from the first function.
+    :param row_key_fields: Fields to use as row key. Defaults to locus.
+    :param interval_ht: Optional table of intervals to filter to.
+    :param entry_keep_fields: Fields to keep in entries before performing the
+        densification in `densify_all_reference_sites`. Should include any fields
+        needed for the functions in `entry_agg_funcs`. By default, only GT or LGT is
+        kept.
+    :param strata_expr: Optional list of dicts of expressions to stratify by.
+    :param group_membership_ht: Optional Table of group membership annotations.
+    :return: Table of stats per site.
     """
     is_vds = isinstance(mtds, hl.vds.VariantDataset)
     if is_vds:
         mt = mtds.variant_data
     else:
         mt = mtds
 
-    # Determine the genotype field.
-    gt_field = set(mt.entry) & {"GT", "LGT"}
-    if len(gt_field) == 0:
-        raise ValueError("No genotype field found in entry fields.")
-
-    gt_field = gt_field.pop()
-
     if group_membership_ht is not None and strata_expr is not None:
         raise ValueError(
             "Only one of 'group_membership_ht' or 'strata_expr' can be specified."
         )
 
+    # Determine if the adj annotation is needed. It is only needed if "adj_groups" is
+    # in the globals of the group_membership_ht and any entry is True, or "freq_meta"
+    # is in the globals of the group_membership_ht and any entry has "group" == "adj".
+    g = {} if group_membership_ht is None else group_membership_ht.globals
+    adj = hl.eval(
+        hl.any(g.get("adj_groups", hl.empty_array("bool")))
+        | hl.any(
+            g.get("freq_meta", hl.empty_array("dict<str, str>")).map(
+                lambda x: x.get("group", "NA") == "adj"
+            )
+        )
+    )
+
+    # Determine the entry fields on mt that should be densified.
```

Comment:
OK, I fixed this to be what I actually wanted here. Here is some testing:

```
def test_fields(en, adj, entry_keep_fields):
    print(f"\nTesting: en: {en}, adj: {adj}, entry_keep_fields: {entry_keep_fields}")
    gt_field = (en & {"GT"} or en & {"LGT"})
    ad_field = (en & {"AD"} or en & {"LAD"})
    adj_fields = en & {"adj"} or ({"DP", "GQ"} | ad_field) if adj else set([])

    if not gt_field:
        print("\tNo genotype field found in entry fields!")
        return

    if adj and not adj_fields.issubset(en):
        print(
            "\tNo 'adj' found in entry fields, and one of AD/LAD, DP, and GQ is missing "
            "so adj can't be computed!"
        )
        return

    entry_keep_fields = set(entry_keep_fields or set([])) | gt_field | adj_fields
    
    print("\tgt_field", gt_field)
    print("\tad_field", ad_field)
    print("\tadj_fields", adj_fields)
    print("\tentry_keep_fields", entry_keep_fields)
    

test_fields(en=set([]), adj=False, entry_keep_fields=set([]))
test_fields(en={"GT"}, adj=False, entry_keep_fields=set([]))
test_fields(en={"GT", "AD", "adj"}, adj=False, entry_keep_fields=set([]))

test_fields(en=set([]), adj=True, entry_keep_fields=set([]))
test_fields(en={"GT"}, adj=True, entry_keep_fields=set([]))
test_fields(en={"GT", "adj"}, adj=True, entry_keep_fields=set([]))
test_fields(en={"GT", "AD"}, adj=True, entry_keep_fields=set([]))
test_fields(en={"GT", "adj", "AD"}, adj=True, entry_keep_fields=set([]))
test_fields(en={"GT", "adj", "DP", "GQ"}, adj=True, entry_keep_fields=set([]))
test_fields(en={"GT", "AD", "DP", "GQ"}, adj=True, entry_keep_fields=set([]))
test_fields(en={"GT", "adj", "AD", "DP", "GQ"}, adj=True, entry_keep_fields=set([]))
```

I switched GT to this to be consistent with what I needed for AD. And I needed AD to be like this because I needed it to only choose one field in case (though unlikely) there was both AD and LAD, I only need one for the set below, and this way since I don't need to keep the actual value, I don't need the pop
    gt_field = en & {"GT"} or en & {"LGT"}
    ad_field = en & {"AD"} or en & {"LAD"}

Reply:
Ah Ok, that makes a lot of sense, thanks for that testing too!